docs: initial Diataxis site (#248)
jnsgruk/concierge: docs/how-to/provide-credentials.md

---
myst:
  html_meta:
    description: How to give Concierge the credentials it needs to bootstrap Juju controllers on clouds like Google, AWS, and Azure using the credentials-file field.
---

(how-to-provide-credentials)=

> # Provide cloud credentials

Reviewer: We ought to add "How to" to the page title (but not include this in the navigation title)

```suggestion
# How to provide cloud credentials
```

Some Juju providers — such as Google, AWS, and Azure — require credentials before Concierge can bootstrap a controller on them. Concierge accepts these through the `credentials-file` field on the provider.

> Providers with built-in Juju credentials (LXD, MicroK8s, K8s) do not need this.

Reviewer: ```suggestion
You don't need to provide credentials for LXD, Canonical Kubernetes, or MicroK8s.
```

## Expected file format

> Concierge expects the file to contain **only** the credential body, without the surrounding `credentials:` / cloud / credential-name keys that Juju uses in `~/.local/share/juju/credentials.yaml`.

Reviewer: I found

```
`credentials:` / cloud / credential-name keys
```

hard to parse. Is there a different way to express this?

Author: ```suggestion
Concierge expects the file to contain **only** the credential body, without the three enclosing keys (`credentials:`, the cloud name, and the credential name) that Juju uses in `~/.local/share/juju/credentials.yaml`.
```

What about this? It loses a bit of detail but has what's important.

Reviewer: I find that better, yes. That makes it clearer we're referring to the key hierarchy.

For example, a Google credential file:

```yaml
auth-type: oauth2
client-email: juju-gce-1-sa@example.iam.gserviceaccount.com
client-id: "1234567891234"
private-key: |
  -----BEGIN PRIVATE KEY-----
  deadbeef
  -----END PRIVATE KEY-----
project-id: example
```

> ## Extract from existing Juju credentials

Reviewer: ```suggestion
## Extract credentials from Juju
```

If you already have credentials in `~/.local/share/juju/credentials.yaml`, extract the block you need with `yq`:

```bash
yq -r '.credentials.google.mycred' \
  ~/.local/share/juju/credentials.yaml > google-creds.yaml
```

## Reference the file from your config

```yaml
providers:
  google:
    enable: true
    bootstrap: true
    credentials-file: /home/ubuntu/google-creds.yaml
```

You can also point Concierge at the file from the command line:

```bash
sudo concierge prepare -c concierge.yaml \
  --google-credential-file /home/ubuntu/google-creds.yaml
```

Or through an environment variable:

```bash
export CONCIERGE_GOOGLE_CREDENTIAL_FILE=/home/ubuntu/google-creds.yaml
sudo concierge prepare -c concierge.yaml

> ```

Reviewer: ````suggestion
## Provide credentials to Concierge

In your config:

```yaml
providers:
  google:
    enable: true
    bootstrap: true
    credentials-file: /home/ubuntu/google-creds.yaml
```

Or on the command line:

```bash
sudo concierge prepare -c concierge.yaml \
  --google-credential-file /home/ubuntu/google-creds.yaml
```

Or through an environment variable:

```bash
export CONCIERGE_GOOGLE_CREDENTIAL_FILE=/home/ubuntu/google-creds.yaml
sudo concierge prepare -c concierge.yaml
```
````
